refactor: share blockmetadata with BlockManager

jamesob opened the pull request:
This is part of the [assumeutxo project](https://github.com/bitcoin/bitcoin/projects/11):

Parent PR: #15606 
Issue: #15605 
Specification: https://github.com/jamesob/assumeutxo-docs/tree/2019-04-proposal/proposal

---

Under an assumeutxo model, we have multiple CChainState instances in use at once in order to support background validation. Currently, each CChainState instance has its own mapBlockIndex, a collection of linked block headers, in addition to a few other data structures that are related to maintenance of the block tree but not necessarily to any given chainstate.

In order to avoid duplicating this data across chainstates, this change moves chainstate-agnostic block metadata (and related behavior) into a class, `BlockManager`. Chainstates are parameterized with a reference to a blockmanager instance and in practice they share the same instance.

Most of this change is conceptually move-only, though the diff is somewhat muddled. The first commit can be reviewed slightly more easily with `--color-moved=dimmed_zebra`. Admittedly, that commit is pretty unwieldy; I tried to split it up after the fact with `git add --patch`, but that was difficult because of git's inability to split hunks past a certain point. Some of the moves also ended up being obscured when done over separate commits.

jamesob committed:
refactoring: move block metadata structures into BlockManager

Separate out the management of chain-agnostic block metadata from any given
CChainState instance. This allows us to avoid duplicating data like
`mapBlockIndex` unnecessarily for multiple chainstates.

This also adds a CChainState constructor that accepts and sets m_blockman.
Ultimately this reference will point to a BlockMan instance that
is shared across CChainStates.

This commit can be decomposed into smaller commits if necessary.

jamesob committed:
refactoring: remove mapBlockIndex global

in lieu of ::BlockIndex().

ryanofsky commented:
Concept ACK

> that was difficult because of git's inability to split hunks past a certain point

First commit seems reasonable to me, and I wouldn't see a need to split it up further. But just for reference I'd recommend trying git add --patch's split and edit commands if you haven't used them before. The split command is good enough for splitting up changes most of the time, and when it isn't, the edit command is fallback that works really well.

```
s - split the current hunk into smaller hunks
e - manually edit the current hunk
```

promag reviewed (commented):
Concept ACK another man to the gang. Overall changes look good to me but will look closely.

promag commented on src/validation.cpp:
```diff
@@ -77,7 +77,9 @@ bool CBlockIndexWorkComparator::operator()(const CBlockIndex *pa, const CBlockIn
 }
 
-CChainState g_chainstate;
+BlockManager g_blockman;
```
7064f3045f41731dd8044cb6ef4277daedb06a2c

static or in a anonymous namespace?

promag commented on src/validation.h:
```diff
@@ -571,23 +610,23 @@ class CChainState {
+    bool IsInitialBlockDownload() const;
+
+    //! @returns the map of blocks that this chainstate is aware of.
+    BlockMap& BlockIndex();
```
Can be const?
```cpp
BlockMap& BlockIndex() const;
```

jamesob commented:
Thanks for the look, guys. I've made @promag's suggested changes and squashed the second commit into the first.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -77,7 +77,11 @@ bool CBlockIndexWorkComparator::operator()(const CBlockIndex *pa, const CBlockIn
+    BlockManager g_blockman;
+} // anon namespace
+
+CChainState g_chainstate(g_blockman);
```
Should be static

MarcoFalke commented on src/validation.cpp:
```diff
@@ -126,12 +130,7 @@ CScript COINBASE_FLAGS;
-     * Pruned nodes may have entries where B is missing data.
-     */
-    std::multimap<CBlockIndex*, CBlockIndex*>& mapBlocksUnlinked = ::ChainstateActive().mapBlocksUnlinked;
+    CBlockIndex* pindexBestInvalid = nullptr;
```
Why is this changed?

jamesob replied:
Ah sorry, this was a little confusing. Because `pindexBestInvalid` is only used in validation.cpp, I removed it as a member of `CChainState` (though in this revision I forgot to actually remove the declaration - willfix) since it doesn't need to be a part of that interface. There is a change in name only, because `::ChainstateActive().pindexBestInvalid` is nullptr at the time of this declaration.

jamesob replied:
Also worth pointing out that we need to touch this piece of data because it is used by both `CChainState` methods *and* `BlockManager` methods.

jamesob replied:
> pindexBestInvalid just seems like it should be a BlockManager member. Or am I missing something?

In theory we could, the only problem here is that (i) [`g_blockman` is retired in later commits](https://github.com/bitcoin/bitcoin/pull/15606/commits/df81dc462e72047ed6857d927946cb18e0e9492b#diff-24efdb00bfbe56b140fb006b562cc70bL81) and (ii) pindexBestInvalid is used in non-method functions like `CheckForkWarningConditions`, so essentially there'd be no way to reference `m_blockman.pindexBestInvalid` in functions like that without moving them onto CChainState. So instead of creating extra diff, I just figured keeping pindexBestInvalid private to validation.cpp made sense. Let me know if you think it'd be better as a member of BlockManager.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/16194#discussion_r296840549

Thanks for explaining. It would be nice to clean up and document these free floating global variables (pindexBestInvalid, pindexBestForkTip, pindexBestForkBase), giving them `g_` prefixes or moving them to appropriate classes so they don't look like local variables. It's beyond the scope of this PR, though.

FWIW, I don't see later commits interfering here. It looks like there still is a single blockman instance, just called g_chainman.m_blockman instead of g_blockman.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -3521,16 +3521,18 @@ void PruneOneBlockFile(const int fileNumber)
+            // m_blocks_unlinked or setBlockIndexCandidates.
+            std::pair<std::multimap<CBlockIndex*, CBlockIndex*>::iterator,
+                std::multimap<CBlockIndex*, CBlockIndex*>::iterator> range =
+                    g_blockman.m_blocks_unlinked.equal_range(pindex->pprev);
```
Could use `const auto` instead of splitting this into three lines.

jamesob replied:
Happy to do that, but thought that the explicit type declaration might be helpful since the value is so complicated. If you think `const auto` is better I'll change it.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -4049,10 +4051,12 @@ void CChainState::EraseBlockData(CBlockIndex* index)
-    std::pair<std::multimap<CBlockIndex*, CBlockIndex*>::iterator, std::multimap<CBlockIndex*, CBlockIndex*>::iterator> ret = mapBlocksUnlinked.equal_range(index->pprev);
+    std::pair<std::multimap<CBlockIndex*, CBlockIndex*>::iterator,
+        std::multimap<CBlockIndex*, CBlockIndex*>::iterator> ret =
+            m_blockman.m_blocks_unlinked.equal_range(index->pprev);
```
Same

MarcoFalke commented on src/validation.h:
```diff
@@ -441,27 +441,88 @@ struct CBlockIndexWorkComparator
+            delete entry.second;
+        }
+
+        m_block_index.clear();
```
Implementation should probably stay in the cpp file, no?

MarcoFalke commented on src/validation.h:
```diff
@@ -509,15 +550,24 @@ class CChainState {
     mutable std::atomic<bool> m_cached_finished_ibd{false};
 
+    //! Reference to a BlockManager instance which itself is shared across all
+    //! CChainState instances.
```
Could explain why this is useful to use instead of the g_blockman.

MarcoFalke commented on src/validation.h:
```diff
@@ -571,23 +616,20 @@ class CChainState {
     void CheckBlockIndex(const Consensus::Params& consensusParams);
 
+    /** Check whether we are doing an initial block download (synchronizing from disk or network) */
+    bool IsInitialBlockDownload() const;
```
Why is this moved? Looks like this is `public` before and after.

jamesob replied:
Good catch, not sure how that got moved.

maflcko reviewed (commented):
Some style questions

jamesob commented:
Rebased and incorporated @MarcoFalke's feedback

maflcko reviewed (commented):
Just two dev-doc-nits

ryanofsky commented on src/validation.cpp:
```diff
@@ -3734,7 +3734,7 @@ bool CChainState::LoadBlockIndex(const Consensus::Params& consensus_params, CBlo
         }
         if (pindex->IsValid(BLOCK_VALID_TRANSACTIONS) && (pindex->HaveTxsDownloaded() || pindex->pprev == nullptr))
-            setBlockIndexCandidates.insert(pindex);
+            ::ChainstateActive().setBlockIndexCandidates.insert(pindex);
```
In commit "refactoring: move block metadata structures into BlockManager" (3a8f92e51f60287492d1a1bce5e5e58ae029bb67)

This BlockManager -> ChainstateActive() dependency seems like it will make BlockManager harder to test independently and have confusing side effects. Would suggest reverting this line and adding an explicit `std::set<CBlockIndex*, CBlockIndexWorkComparator>& setBlockIndexCandidates` output parameter to this function. It's only called one place, and I think it'd make the call site clearer, because otherwise you might not know calling this has any effect on the active chain.

ryanofsky commented on src/validation.cpp:
```diff
@@ -1045,6 +1051,11 @@ bool CChainState::IsInitialBlockDownload() const
 
 static CBlockIndex *pindexBestForkTip = nullptr, *pindexBestForkBase = nullptr;
 
+BlockMap& CChainState::BlockIndex() const
```
In commit "refactoring: remove mapBlockIndex global" (1e33f16dd47c42a9f6099f6172ebd5f8fbb32ec8)

I'm probably just bikeshedding, but I think having this method is counterproductive and only obfuscates what's going on. In places where this is called in `CChainState` methods it seems clearer to replace `BlockIndex()` calls with direct `m_blockman.m_block_index` references. In places where this is called globally, it would seem clearer to able to just write `::BlockIndex()` instead of `::ChainstateActive().BlockIndex()` since the block index doesn't depend on the active chain, and because it would be more consistent with the existing `::LookupBlockIndex` calls.

jamesob replied:
Yep, I agree with this. Willfix.

ryanofsky reviewed (approved):
utACK 1e33f16dd47c42a9f6099f6172ebd5f8fbb32ec8. Code changes all look correct. Left some suggestions and feedback, but feel free to ignore it.

jamesob commented on src/validation.cpp:
```diff
@@ -3304,7 +3304,7 @@ bool CChainState::AcceptBlockHeader(const CBlockHeader& block, CValidationState&
         *ppindex = pindex;
 
-    CheckBlockIndex(chainparams.GetConsensus());
+    ::ChainstateActive().CheckBlockIndex(chainparams.GetConsensus());
```
Yep, agree. Willfix.

jamesob committed:
refactoring: make pindexBestInvalid internal to validation.cpp

There's no need to have this member live on CChainState since it's only used
in validation.cpp.

jamesob committed:
refactoring: add block_index_candidates arg to LoadBlockIndex

Prevents BlockManager from having to reference ChainstateActive()
within one of its methods which improves encapsulation and makes
testing easier.

jamesob commented:
Thanks for the good feedback, @MarcoFalke @ryanofsky. I've incorporated your changes.

jamesob commented on src/validation.cpp:
```diff
@@ -3300,8 +3311,6 @@ bool CChainState::AcceptBlockHeader(const CBlockHeader& block, CValidationState&
     if (ppindex)
         *ppindex = pindex;
 
-    CheckBlockIndex(chainparams.GetConsensus());
```
Note to reviewers: this is moved to the two callsites below to avoid making reference to a specific CChainState in BlockManager.

MarcoFalke replied:
If this refactoring seems fragile in light of silent merge conflicts, an alternative would be to pass in a reference to the chainstate and then call CBI on that reference. This way, a new call to `AcceptBlockHeader` introduced before your change will not be a silent merge conflict if merged with your chagnes.

Sjors replied:
As @sdaftuar pointed out in https://github.com/bitcoin/bitcoin/issues/16444#issuecomment-514801708 `AcceptBlockHeader` can return early, so we're now calling `CheckBlockIndex(()` more often than before.

ryanofsky reviewed (approved):
utACK e89e4f62b8cf4d19becb66f0c325d40851040a35. One changes since last review were various review suggestions discussed above, and splitting some small parts of the first commit into separate commits.

ryanofsky reviewed (approved):
utACK 682a1d0f2004d808b87b3106d0dfae547005e638, only changes since last review were rebase and fixing conflict on a moved line

MarcoFalke commented:
ACK 682a1d0f2004d808b87b3106d0dfae547005e638

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 682a1d0f2004d808b87b3106d0dfae547005e638
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhwngwAq5VhfgfOv6FJFWvZc8NpFrYVMPW/S6u8qLBlZs3q2KZMT7j12l0MQRmX
qwGyP+CRJKsuJvYIwShRGHV5el/0t5hkTD8HUMlGgXSSkQvuEJCCweD2t1Y1y8Pm
Ayi3QErIRcS1ErtnP7dxgBSjm8DOeZJdv+hKehG8kfwjAYePc+rEQhL84nT6wTgU
qNkk7k3ri8eSce6ZEy/3aOHONmHoTw7iqdyqQaUpuqvBSzig5HZ09wx9BR7lic8Z
yAyGmyaYBsfnyl6sl7d+u7EC8UjMQe13Tj9dW+rLYX9vCJz5HSLc/D4SioErNFe1
gcHhk0E7bkLxjGvxO8vMOjRwDR/CCX0ubWYhXliAqNV8cExmr6/Z9aPHVa2oj7KL
j7JBf1sH8qczMDlZqh+/qbN4i9/Ss7oiFVhALAUST07bMs3m9RdbyQ4Fuw6IqoI0
daL3BadGheJmLdAeI1lKXFiaqQp8HNiMi02210gkhmVkHYI8zFSEXGSLt+xuySnw
O0MY7j8+
=6aAq
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `5b30b7a598d04185a0efdbef0b0f6e39648bf992b0c23151e41ad9445e741517  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401085b30b7a598d04185a0efdbef0b0f6e39648bf992b0c23151e41ad9445e741517f0101143d65c7c0e5e4ee14a3bb212045ebc08fff01071973946bb09d0d50331ab4c81aa29fc08f120008c6ab34a032d2620d27c1dc8a61f5a6897c4187a2c04fb6ff52c69acae364108f1045d28aaeaf00840dce6964d3f28ec0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010d274d8d47e451d97ee51146b17d0b26208f0207b1b9c869ed8e9025bb8e8ce6d779075e4f6e4805ff279a01e19a7e0c7e9d34808f1045d28aaeaf008ff3a553406e1a56f0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010dbfd683ecbb21646a0e4f2badd857fff08f0205408c61f4651fc8e8dd386d826d5ebf9032a77c614d39ad7e9a8de856e12ee9908f1045d28aaeaf008dbde88c3326eaff40083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010e9427864b2d8126eebe65f71a45f5ac208f0203ad63966517fc0f8493ba13fc444d6d7186f14565c84b31321f3c1f6d59965a908f1045d28aaeaf008ca9d9a89449f45830083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

ariard commented on src/validation.h:
```diff
@@ -475,9 +475,19 @@ class BlockManager {
+    /**
+     * Load the blocktree off disk and into memory. Populate certain metadata
+     * per index entry (nStatus, nChainWork, nTimeMax, etc.) as well as peripheral
+     * collections like setDirtyBlockIndex.
```
hmmm maybe setDirtyBlockIndex could be part of BlockManager in future commits given this set of block index  is also shared between chainstates? And would let implement batching managed by BlockManager

ariard commented on src/validation.cpp:
```diff
@@ -147,6 +146,13 @@ namespace {
     std::set<int> setDirtyFileInfo;
 } // anon namespace
 
+CBlockIndex* LookupBlockIndex(const uint256& hash)
```
It's unclear reading commits messages what's the long term destination of `g_blockman` is. Do other parts of the code calling `LookupBlockIndex` will always rely on a global `BlockManager` or should they access to it through their instance of `CChainstate`  after future refactoring ?

jamesob replied:
Eventually, a single `BlockManager` object will be managed and shared to all created `CChainState` objects by something called `ChainstateManager`, to be introduced later (https://github.com/bitcoin/bitcoin/pull/15606/commits/df81dc462e72047ed6857d927946cb18e0e9492b). The single global block index will be accessed through the chainstate manager.

ariard reviewed (approved):
utACK 682a1d0. Most of the changes are move-only, with main problem being to avoid creating circular dependencies between `BlockManager` and `CChainState`. Tested, comments are mostly nits, feel free to ignore them

jamesob commented on src/validation.h:
```diff
@@ -439,27 +439,80 @@ struct CBlockIndexWorkComparator
+    std::set<CBlockIndex*> m_failed_blocks;
+
+    /**
+     * All pairs A->B, where A (or one of its ancestors) misses transactions, but B has transactions.
```
Agree, but don't want to invalidate the move-only nature for this change. Would be a great follow-up if you're on the hunt for PR ideas :).

jamesob commented on src/validation.h:
```diff
@@ -439,27 +439,80 @@ struct CBlockIndexWorkComparator
+    /** Clear all data members. */
+    void Unload() EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+
+    CBlockIndex* AddToBlockIndex(const CBlockHeader& block) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
Good idea - if I have to rebase this PR for some other reason I'll add it.

rebroad commented on src/validation.cpp:
```diff
@@ -3319,7 +3328,10 @@ bool ProcessNewBlockHeaders(const std::vector<CBlockHeader>& headers, CValidatio
             CBlockIndex *pindex = nullptr; // Use a temp pindex instead of ppindex to avoid a const_cast
-            if (!::ChainstateActive().AcceptBlockHeader(header, state, chainparams, &pindex)) {
+            bool accepted = g_blockman.AcceptBlockHeader(header, state, chainparams, &pindex);
+            ::ChainstateActive().CheckBlockIndex(chainparams.GetConsensus());
```
is this necessary when accepted is false?

MarcoFalke replied:
#16453

rebroad commented on src/validation.cpp:
```diff
@@ -3362,7 +3374,10 @@ bool CChainState::AcceptBlock(const std::shared_ptr<const CBlock>& pblock, CVali
 
-    if (!AcceptBlockHeader(block, state, chainparams, &pindex))
+    bool accepted_header = m_blockman.AcceptBlockHeader(block, state, chainparams, &pindex);
+    CheckBlockIndex(chainparams.GetConsensus());
```
is this necessary when accepted_header is false?

MarcoFalke replied:
#16453
